src/mlpack/methods/ann/layer/gru_impl.hpp

```diff
@@ -0,0 +1,279 @@
+  network.push_back(forgetGateModule);
+
+  prevOutput = arma::zeros<arma::mat>(outSize, 1);
+  prevError = arma::zeros<arma::mat>(3 * outSize, 1);
+
+  outParameter.reserve(rho);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void GRU<InputDataType, OutputDataType>::Forward(
+    arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  if (!deterministic)
+  {
+    outParameter.push_back(prevOutput);
+  }
+  
+  // process the input linearly(zt, rt, ot)
+  boost::apply_visitor(ForwardVisitor(std::move(input), std::move(
+      boost::apply_visitor(outputParameterVisitor, input2GateModule))),
+      input2GateModule);
+
+  // process the output(zt, rt) linearly
+  boost::apply_visitor(ForwardVisitor(std::move(prevOutput), std::move(
+      boost::apply_visitor(outputParameterVisitor, output2GateModule))),
+      output2GateModule);
+      
+  // merge the outputs(zt and rt)
+  output = (boost::apply_visitor(outputParameterVisitor, 
+      input2GateModule).submat(0, 0, 2 * outSize - 1, 0) +
+      boost::apply_visitor(outputParameterVisitor, output2GateModule));
+
+  // pass the first outSize through inputGate
+  boost::apply_visitor(ForwardVisitor(std::move(output.submat(
+      0, 0, 1 * outSize - 1, 0)), std::move(boost::apply_visitor(
+      outputParameterVisitor, inputGateModule))), inputGateModule);
+      
+  // pass the second through forgetGate
+  boost::apply_visitor(ForwardVisitor(std::move(output.submat(
+      1 * outSize, 0, 2 * outSize - 1, 0)), std::move(boost::apply_visitor(
+      outputParameterVisitor, forgetGateModule))), forgetGateModule);
+      
+  //temp = output of forgetGate(rt) % prevOutput
+  arma::mat modInput = (boost::apply_visitor(outputParameterVisitor, 
+      forgetGateModule) % prevOutput);
+      
+  // pass that through the outputHidden2GateModule
+  boost::apply_visitor(ForwardVisitor(std::move(modInput), std::move(
+      boost::apply_visitor(outputParameterVisitor, outputHidden2GateModule))),
+      outputHidden2GateModule);
+  
+  // merge them
+  arma::mat outputH = boost::apply_visitor(outputParameterVisitor, 
+      input2GateModule).submat(2 * outSize, 0, 3 * outSize - 1, 0) + 
+      boost::apply_visitor(outputParameterVisitor, outputHidden2GateModule);
+  
+  // pass it through hiddenGate
+  boost::apply_visitor(ForwardVisitor(std::move(outputH), std::move(
+      boost::apply_visitor(outputParameterVisitor, hiddenStateModule))), 
+      hiddenStateModule);
+
+  // Update the output (nextOutput): cmul1 + cmul2
+  // where cmul1 is input gate * prevOutput and
+  // cmul2 is (1 - input gate) * hidden gate.
+  output = (boost::apply_visitor(outputParameterVisitor, inputGateModule) 
+      % prevOutput) + 
+      ((arma::ones<arma::vec>(outSize) - 
+      boost::apply_visitor(outputParameterVisitor, inputGateModule)) % 
+      boost::apply_visitor(outputParameterVisitor,
+      hiddenStateModule));
+      
+  prevOutput = output;
```

Comment:
Hmm... I agree. We could use the last entry in outParameters rather than prevOutput, right? or am I getting it wrong? I think it will save 2 copies. If we do push the temporary computed output directly to outParameters with c++11 move semantics and then make a copy of that for returning. I think we can do the same thing for LSTM as well. What do you think?